cockroachdb/resources/strobetime.c

```diff
@@ -0,0 +1,156 @@
+#include <stdio.h>
+#include <time.h>
+#include <sys/time.h>
+#include <stdlib.h>
+#include <stdint.h>
+
+const int64_t NANOS_PER_SEC = 1000000000;
+
+# define TIMEVAL_TO_TIMESPEC(tv, ts) {                                   \
+    (ts)->tv_sec = (tv)->tv_sec;                                    \
+    (ts)->tv_nsec = (tv)->tv_usec * 1000;                           \
+}
+# define TIMESPEC_TO_TIMEVAL(tv, ts) {                                   \
+    (tv)->tv_sec = (ts)->tv_sec;                                    \
+    (tv)->tv_usec = (ts)->tv_nsec / 1000;                           \
+}
+
+/* Convert nanoseconds to a timespec */
+struct timespec nanos_to_timespec(int64_t nanos) {
+  struct timespec t;
+  int64_t dnanos   = nanos % NANOS_PER_SEC;
+  int64_t dseconds = (nanos - dnanos) / NANOS_PER_SEC;
+  t.tv_nsec = dnanos;
+  t.tv_sec = dseconds;
+  return t;
+}
+
+/* Obtain monotonic clock as a timespec */
+struct timespec monotonic_now() {
+  struct timespec now;
+  clock_gettime(CLOCK_MONOTONIC, &now);
+  return now;
+}
+
+/* Obtain wall clock as a timespec */
+struct timespec wall_now() {
+  struct timespec now_ts;
+  struct timeval  now_tv;
+  struct timezone tz;
+  if (0 != gettimeofday(&now_tv, &tz)) {
+    perror("gettimeofday");
+    exit(1);
+  }
+  TIMEVAL_TO_TIMESPEC(&now_tv, &now_ts);
+  return now_ts;
+}
+
+/* Set wall clock */
+void set_wall_clock(struct timespec ts) {
+  struct timeval tv;
+  TIMESPEC_TO_TIMEVAL(&tv, &ts);
+  // printf("Setting clock: %d %d\n", tv.tv_sec, tv.tv_usec);
+  if (0 != settimeofday(&tv, NULL)) {
+    perror("settimeofday");
+    exit(2);
+  }
+}
+
+/* Rebalances sec/nsec to be within bounds. Mutates t.*/
```

Comment:
I don't understand what this function does. Can you explain?

Reply:
oh I see.

Reply:
Btw I dug into this and what it does it ensures that ns is always a fraction of second.
It is called after time adjustments that could make ns component be less than zero or more than second. So it loops up if small until it jumps out of range and then loops down till it gets into the range. So for ns too large first loop is skipped and then second adjusts down. For negative values it would grow 1 sec more than needed and then second loop adjusts down.
Alternative would be to have two branches and use divide and modulo to put ns into range.